golem-cli/src/model/wave.rs

```diff
@@ -0,0 +1,282 @@
+use crate::model::GolemError;
+use golem_client::model::{
+    Export, ExportFunction, ResourceMode, Template, Type, TypeHandle, TypeRecord, TypeResult,
+    TypeTuple, TypeVariant,
+};
+use golem_wasm_ast::analysis::{AnalysedResourceId, AnalysedResourceMode, AnalysedType};
+use std::borrow::Cow;
+use std::fmt::Display;
+use tracing::info;
+use wasm_wave::wasm::{DisplayType, WasmFunc};
+
+pub fn type_wave_compatible(typ: &Type) -> bool {
+    fn variant_wave_compatible(tv: &TypeVariant) -> bool {
+        tv.cases
+            .iter()
+            .all(|notp| notp.typ.iter().all(type_wave_compatible))
+    }
+
+    fn result_wave_compatible(tr: &TypeResult) -> bool {
+        tr.ok.iter().all(type_wave_compatible) && tr.err.iter().all(type_wave_compatible)
+    }
+
+    fn record_wave_compatible(tr: &TypeRecord) -> bool {
+        tr.cases.iter().all(|ntp| type_wave_compatible(&ntp.typ))
+    }
+
+    fn tuple_wave_compatible(tt: &TypeTuple) -> bool {
+        tt.items.iter().all(type_wave_compatible)
+    }
+
+    match typ {
+        Type::Variant(tv) => variant_wave_compatible(tv),
+        Type::Result(tr) => result_wave_compatible(tr),
+        Type::Option(to) => type_wave_compatible(&to.inner),
+        Type::Enum(_) => true,
+        Type::Flags(_) => true,
+        Type::Record(tr) => record_wave_compatible(tr),
+        Type::Tuple(tt) => tuple_wave_compatible(tt),
+        Type::List(tl) => type_wave_compatible(&tl.inner),
+        Type::Str(_) => true,
+        Type::Chr(_) => true,
+        Type::F64(_) => true,
+        Type::F32(_) => true,
+        Type::U64(_) => true,
+        Type::S64(_) => true,
+        Type::U32(_) => true,
+        Type::S32(_) => true,
+        Type::U16(_) => true,
+        Type::S16(_) => true,
+        Type::U8(_) => true,
+        Type::S8(_) => true,
+        Type::Bool(_) => true,
+        Type::Handle(_) => false,
+    }
+}
+
+pub fn function_wave_compatible(func: &ExportFunction) -> bool {
+    func.parameters.iter().all(|p| type_wave_compatible(&p.typ))
+        && func.results.iter().all(|r| type_wave_compatible(&r.typ))
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct WrapExportFunction(pub ExportFunction);
+
+fn wrap_analysed_type(typ: AnalysedType) -> golem_wasm_rpc::AnalysedType {
+    golem_wasm_rpc::AnalysedType(typ)
+}
+
+pub fn wrap_type(typ: &Type) -> golem_wasm_rpc::AnalysedType {
+    wrap_analysed_type(type_to_analysed(typ))
+}
+
+impl WasmFunc for WrapExportFunction {
```

Comment:
Ah I did not know there is a function type class in wasm-wave too! That's nice. 
And we should clean this and wasm-rpc up by implementing these typeclasses directly in `golem-wasm-ast` (under a `wave` feature flag). But that can be a separate task of course.

Reply:
I'll move this into `golem-wasm-rpc` after merging this one.
Please note that `DisplayWasmFunc` in `wasm-wave` is not usable for us - it renders `func` as function name (hardcoded).

Reply:
I think the correct place for the instances is https://github.com/golemcloud/golem-wasm-ast/blob/main/src/analysis/mod.rs where `AnalysedType` and `AnalysedFunction` is defined.

This means removing the wrapper and instance from `wasm-rpc` too.